### PR TITLE
RATIS-2183. Detect staled snapshot request.

### PR DESCRIPTION
**Pull request by 133tosakarin**
see https://issues.apache.org/jira/browse/RATIS-2183

### CODE REVIEW

**Comment by szetszwo** [comment]
@133tosakarin , we already have 
- `callId` in https://github.com/apache/ratis/blob/bf15765a58ac1476a13239d812df0c0b172acb22/ratis-proto/src/main/proto/Raft.proto#L221 
and 

- `termIndex` in https://github.com/apache/ratis/blob/bf15765a58ac1476a13239d812df0c0b172acb22/ratis-proto/src/main/proto/Raft.proto#L211

For new and old snapshotRequest detection, could we reuse them instead of adding a new timestamp?

**Comment by 133tosakarin** [comment]
> @133tosakarin , we already have
> 
> * `callId` in https://github.com/apache/ratis/blob/bf15765a58ac1476a13239d812df0c0b172acb22/ratis-proto/src/main/proto/Raft.proto#L221
>   
>   and
> * `termIndex` in https://github.com/apache/ratis/blob/bf15765a58ac1476a13239d812df0c0b172acb22/ratis-proto/src/main/proto/Raft.proto#L211
> 
> For new and old snapshotRequest detection, could we reuse them instead of adding a new timestamp?

Thanks for the reminder, just use callId.

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

File: ratis-server/src/main/java/org/apache/ratis/server/impl/SnapshotInstallationHandler.java [hunk 1]
```diff
@@ -176,8 +177,16 @@ private CompletableFuture<InstallSnapshotReplyProto> checkAndInstallSnapshot(Ins
       state.setLeader(leaderId, "installSnapshot");
 
       server.updateLastRpcTime(FollowerState.UpdateType.INSTALL_SNAPSHOT_START);
+      if (callId.get() > request.getServerRequest().getCallId()) {
+        LOG.warn("{}: Ignoring stale snapshot request {}", getMemberId(),
+            ServerStringUtils.toInstallSnapshotRequestString(request));
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/SnapshotInstallationHandler.java:182]: Print also the `chunk0CallId`
```java
      final long chunk0 = chunk0CallId.get();
      if (chunk0 > request.getServerRequest().getCallId()) {
        LOG.warn("{}: Snapshot Request Staled: chunk 0 callId is {} but {}", getMemberId(), chunk0,
            ServerStringUtils.toInstallSnapshotRequestString(request));
        final InstallSnapshotReplyProto reply =  toInstallSnapshotReplyProto(leaderId, getMemberId(),
            currentTerm, snapshotChunkRequest.getRequestIndex(), InstallSnapshotResult.SNAPSHOT_EXPIRED);
        return future.thenApply(dummy -> reply);
      }
```

**Review by szetszwo — COMMENTED** [review]
@133tosakarin , thanks for the update!   

- Please revert the white space changes in `LogAppender` and `InstallSnapshotRequests`.
- See also the comments inlined.

File: ratis-server/src/main/java/org/apache/ratis/server/impl/SnapshotInstallationHandler.java [hunk 2]
```diff
@@ -71,6 +71,7 @@ class SnapshotInstallationHandler {
   private final AtomicBoolean isSnapshotNull = new AtomicBoolean();
   private final AtomicLong installedIndex = new AtomicLong(INVALID_LOG_INDEX);
   private final AtomicInteger nextChunkIndex = new AtomicInteger(-1);
+  private final AtomicLong callId = new AtomicLong(-1);
```
[the remark this answers is not in the document]
Comment (133tosakarin) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/SnapshotInstallationHandler.java]: ok

File: ratis-server/src/main/java/org/apache/ratis/server/impl/SnapshotInstallationHandler.java (hunk 1 above)
Comment (133tosakarin) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/SnapshotInstallationHandler.java]: ok

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Comment by szetszwo** [comment]
@133tosakarin , thanks for working on this!

@OneSizeFitsQuorum , thanks also for reviewing this!

**Pull request merged**